fgpyo/util/metric.py

```diff
@@ -466,3 +477,101 @@ def asdict(metric: Metric) -> Dict[str, Any]:
+
+    Args:
+        metric_class: A Metric class.
+
+    Returns:
+        A list of fieldnames.
+
+    Raises:
+        TypeError: If the given class is not a Metric.
+    """
+    _assert_is_metric_class(metric_class)
+
+    if dataclasses.is_dataclass(metric_class):
+        return [f.name for f in dataclasses.fields(metric_class)]
+    elif attr.has(metric_class):
+        return [f.name for f in attr.fields(metric_class)]
+    else:
+        assert False, "Unreachable"
+
+
+def _assert_file_header_matches_metric(
+    path: Path,
+    metric_class: Type[MetricType],
+    delimiter: str,
+) -> None:
+    """
+    Check that the specified file has a header and its fields match those of the provided Metric.
+
+    Args:
+        path: A path to a `Metric` file.
+        metric_class: The `Metric` class to validate against.
+        delimiter: The delimiter to use when reading the header.
+
+    Raises:
+        ValueError: If the provided file does not include a header.
+        ValueError: If the header of the provided file does not match the provided Metric.
+    """
+    _assert_is_metric_class(metric_class)
+
+    header: MetricFileHeader
+    with path.open("r") as fin:
+        try:
+            header = metric_class._read_header(fin, delimiter=delimiter)
+        except ValueError:
+            raise ValueError(f"Could not find a header in the provided file: {path}")
+
+    fieldnames: List[str] = _get_fieldnames(metric_class)
+    if header.fieldnames != fieldnames:
+        raise ValueError(
+            "The provided file does not have the same field names as the provided Metric:\n"
+            f"\tMetric: {metric_class.__name__}\n"
+            f"\tFile: {path}\n"
+            f"\tExpected fields: {', '.join(fieldnames)}\n"
+            f"\tActual fields: {', '.join(header.fieldnames)}\n"
+        )
+
+
+def _assert_fieldnames_are_metric_attributes(
+    specified_fieldnames: List[str],
+    metric_class: Type[MetricType],
+) -> None:
+    """
+    Check that all of the specified fields are attributes on the given Metric.
+
+    Raises:
+        ValueError: if any of the specified fieldnames are not an attribute on the given Metric.
+    """
+    _assert_is_metric_class(metric_class)
+
+    invalid_fieldnames = [
+        f for f in specified_fieldnames if f not in _get_fieldnames(metric_class)
+    ]
+
+    if len(invalid_fieldnames) > 0:
+        raise ValueError(
+            "One or more of the specified fields are not attributes on the Metric "
+            + f"{metric_class.__name__}: "
+            + ", ".join(invalid_fieldnames)
+        )
+
+
+def _assert_is_metric_class(cls: Type[Metric]) -> None:
+    """
+    Assert that the given class is a Metric.
+
+    Args:
+        cls: A class object.
+
+    Raises:
+        TypeError: If the given class is not a Metric.
+    """
+    if not _is_metric_class(cls):
+        raise TypeError(f"Not a dataclass or attr decorated Metric: {cls}")
```

Comment:
How do we feel about `_assert` functions raising errors other than `AssertionError`?  I would probably be surprised by this.

Reply:
I don't think `assert` should be used outside of unit tests, and I prefer to raise more specific exceptions.

The Google style guide [recommends/requires this as well](https://google.github.io/styleguide/pyguide.html#244-decision):
> Do not use assert statements in place of conditionals or validating preconditions. They must not be critical to the application logic. A litmus test would be that the assert could be removed without breaking the code. assert conditionals are [not guaranteed](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement) to be evaluated. For [pytest](https://pytest.org/) based tests, assert is okay and expected to verify expectations.


The primary reason to avoid the use of assert statements is that [they are intended for debugging and can be disabled](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement).

Additionally, using `assert` makes it easy to fail to cover all branches of a program:
```python
# Covered by any unit test that calls the parent function
assert is_ok(foo), "Foo was not ok"

if not is_ok(foo):
    # Not covered unless the condition where `is_ok(foo)` is False is tested explicitly
    raise ValueError("Foo was not ok")  
```

I'm happy to rename the functions to something other than `assert_*` (I've used `validate_*` in the past), I was just trying to remain consistent with naming convention in the project.

Reply:
This might be a good case for having a custom exception:

```python
class ValueAssertion(AssertionError, ValueError):
    pass
```

Then you would catch the exception with either `AssertionError` or `ValueError`.

I do like having the explicit branch with `if` since code coverage analysis will identify it as a covered/uncovered branch depending on actual test coverage. With single-line statements (e.g. `assert`), you don't get any branch information so it is easy to forget to test the exception paths.